File: core/src/main/java/pcb/uwu/amount/composite/fundamental/Frequency.java
package pcb.uwu.amount.composite.fundamental;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.fundamental.FrequencyUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Frequency extends CompositeUnitAmount<FrequencyUnit> {

	// region constructors

	public Frequency(Number value, FrequencyUnit unit) {
		super(value, unit);
	}

	public Frequency(Number value, Magnitude magnitude, FrequencyUnit unit) {
		super(value, magnitude, unit);
	}

	public Frequency(String value, FrequencyUnit unit) {
		super(value, unit);
	}

	public Frequency(String value, Magnitude magnitude, FrequencyUnit unit) {
		super(value, magnitude, unit);
	}

	public Frequency(BigDecimal value, FrequencyUnit unit) {
		super(value, unit);
	}

	public Frequency(BigDecimal value, Magnitude magnitude, FrequencyUnit unit) {
		super(value, magnitude, unit);
	}

	public Frequency(BigDecimalAmount amount, FrequencyUnit unit) {
		super(amount, unit);
	}

	public Frequency(BigDecimalAmount amount, Magnitude magnitude, FrequencyUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Frequency plus(UnitAmount<FrequencyUnit> other, MathContext mathContext) {
		return new Frequency(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public Frequency minus(UnitAmount<FrequencyUnit> other, MathContext mathContext) {
		return new Frequency(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public Frequency multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Frequency(multipliedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Frequency dividedBy(BigDecimal other, MathContext mathContext) {
		return new Frequency(dividedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Frequency convertTo(FrequencyUnit unit) {
		return new Frequency(getAmountIn(unit), unit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/derived/acoustics/Frequency.kt
package pcb.uwu.amount.derived.acoustics

import pcb.uwu.amount.base.Time
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.acoustics.FrequencyUnit
import pcb.uwu.util.UnitAmountUtils.getAmountIn

open class Frequency : CompositeUnitAmount<FrequencyUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: FrequencyUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: FrequencyUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(frequency: UnitAmount<FrequencyUnit>) =
        Frequency(amount = this.amount + (frequency to this.unit).amount,
                  unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(frequency: UnitAmount<FrequencyUnit>) =
        Frequency(amount = this.amount - (frequency to this.unit).amount,
                  unit = this.unit)

    override fun times(scalar: Number) =
        Frequency(amount = this.amount * scalar,
                  unit = this.unit)

    override fun div(scalar: Number) =
        Frequency(amount = this.amount / scalar,
                  unit = this.unit)

    override fun to(unit: FrequencyUnit) =
        Frequency(amount = getAmountIn(unitAmount = this, newUnit = unit),
                  unit = unit)

    override fun invert() =
        Time(amount = amount.invert(),
             unit = unit.unitCounter.findUnit(TimeUnit::class)!!)

    // endregion
}
